### PR TITLE
[XC] Add file info to exception reports

### LINKED ISSUE #21038: XAML errors frequently do not reference the file with an issue
### Description

See title. Sometimes when there's an error with an XAML file, Visual Studio is unable to tell me which file/line the problem occurs in.

I never had this issue with Xamarin Forms, but in one week of Maui it's happened multiple times.

### Steps to Reproduce

I'm not sure if this will reproduce the issue, but with the latest instance, I accidentally added the `Spacing` attribute to a `StackLayout` twice

```
<StackLayout HorizontalOptions="FillAndExpand" Spacing="0" Margin="5,0,5,0" Spacing="6">
    ...
</StackLayout>
```

The error in the error list opened up the .csproj file, rather than the file containing the error.

2024-03-06 01_20_08-NVIDIA GeForce Overlay DT

### Link to public reproduction project repository

_No response_

### Version with bug

8.0.3 GA

### Is this a regression from previous behavior?

Yes, this used to work in Xamarin.Forms

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

_No response_

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
### Description of Change

When the XamlParser throws an exception while executed from XamlC, the file info is missing from the log, and the whole XamlC process fails, so it only reports one issue at a time.

It now logs things like 

```
/Users/sde/Projects/Microsoft/maui/src/Controls/tests/Xaml.UnitTests/Issues/Maui21038.xaml(8,10): XamlC error : 'Label.Text' is a duplicate property name. [/Users/sde/Projects/Microsoft/maui/src/Controls/tests/Xaml.UnitTests/Controls.Xaml.UnitTests.csproj]
```

### Issues Fixed

Fixes #21038

### CODE REVIEW

**Comment by StephaneDelcroix** [comment]
still a draft. I did this while working on something else

**Comment by StephaneDelcroix** [comment]
This is the kind of message logged with this change

```
/Users/sde/Projects/Microsoft/maui/src/Controls/tests/Xaml.UnitTests/Issues/Maui21038.xaml(8,10): XamlC error : 'Label.Text' is a duplicate property name. [/Users/sde/Projects/Microsoft/maui/src/Controls/tests/Xaml.UnitTests/Controls.Xaml.UnitTests.csproj]

File: src/Controls/src/Xaml/XamlParser.cs [hunk 1]
```diff
@@ -76,7 +76,7 @@ static void ParseXamlElementFor(IElementNode node, XmlReader reader)
 								name = new XmlName(reader.NamespaceURI, reader.LocalName);
 
 							if (node.Properties.ContainsKey(name))
-								throw new XamlParseException($"'{reader.Name}' is a duplicate property name.", (IXmlLineInfo)reader);
+								throw new XamlParseException($"'{reader.Name}' is a duplicate property name.", ((IXmlLineInfo)reader).Clone());
```
Comment (StephaneDelcroix) [inline src/Controls/src/Xaml/XamlParser.cs:79]: reader is disposed before the message is logged, so the line info was missing

File: src/Controls/src/Build.Tasks/XamlCTask.cs [hunk 2]
```diff
@@ -243,10 +273,19 @@ public override bool Execute(out IList<Exception> thrownExceptions)
 
 
 							LoggingHelper.LogMessage(Low, $"{new string(' ', 6)}Parsing Xaml");
-							var rootnode = ParseXaml(resource.GetResourceStream(), typeDef);
-							if (rootnode == null)
+							ILRootNode rootnode = null;
+							try {
```
Comment (StephaneDelcroix) [inline src/Controls/src/Build.Tasks/XamlCTask.cs:277]: this try/catch is the main change, the rest is pulling #22808

File: src/Controls/tests/Xaml.UnitTests/Issues/Maui21038.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            DispatcherProvider.SetCurrent(new DispatcherProviderStub());
+        }
+
+        [TearDown] public void TearDown() => AppInfo.SetCurrent(null);
+
+        [Test]
+        public void XamlParseErrorsHaveFileInfo([Values(false, true)] bool useCompiledXaml)
+        {
+            if (useCompiledXaml)            
+                Assert.DoesNotThrow(() => MockCompiler.Compile(typeof(Maui21038)));
```
Comment (StephaneDelcroix) [inline src/Controls/tests/Xaml.UnitTests/Issues/Maui21038.xaml.cs:39]: we don't yet have infrastructure for testing LogMessage. this will come later

**Comment by jsuarezruiz** [comment]
@StephaneDelcroix Could you rebase? Thanks in advance.

**Review by jfversluis — APPROVED** [review]

**Review by rmarinho — APPROVED** [review]

**Pull request merged**